Range-aware Port Per Node with integer configuration

robobario opened the pull request:
### Type of change

- Enhancement / new feature

### Description

Mitigates #902

This new implementation introduces the concept of Node Id Ranges and enables you to configure multiple such ranges for your virtual cluster. Enabling a deterministic mapping of node ids to proxy ports even when there are large gaps between the ranges.

It is implemented as a new ClusterNetworkAddressConfigProvider of type `RangeAwarePortPerNodeClusterNetworkAddressConfigProvider`

configured like:

``` yaml
virtualClusters:
  demo:
    targetCluster:
      bootstrap_servers: localhost:9092
    clusterNetworkAddressConfigProvider:
      type: RangeAwarePortPerNodeClusterNetworkAddressConfigProvider
      config:
        bootstrapAddress: localhost:9192
        nodeIdRanges:
          - name: brokers
            range:
              startInclusive: 0
              endExclusive: 3
```

For example if I configure a virtual cluster with:

```yaml
nodeStartPort: 9000
nodeIdRanges:
  - name: brokers
    range:
      startInclusive: 4
      endExclusive: 6
  - name: controllers
    range:
       startInclusive: 1
       endExclusive: 3
```
Then the ports will be mapped as follows:
- node 1 -> port 9000
- node 2 -> port 9001
- node 4 -> port 9002
- node 5 -> port 9003

Why:
Currently there are numerous drawbacks to using the PortPerBroker exposition scheme. If the lowest broker id is N and the highest is M then we must expose M-N ports so that we can deterministically map each node id to a port. So if you have two nodes with ids 100000 and 200000 you need 100000 ports mapped so that we can say startPort+0 is node 100000 and startPort+100000 is node id 200000.

Also with the advent of KRaft we now may have to contend with two distinct and evolving sets of node ids, a controller pool and a broker pool. If there is a significant gap between them we have the same problem as above.

This enables us to represent clusters with highly variable node ids using a more compact set of ports.

Note it still does not solve other issues, like some users regulurly introduce new nodes and remove old ones in their target cluster as part of upgrades.  Leading to eternal growth of node ids. To handle this in a better way we think the Proxy would need to be able to discover the target topology.

This is an alternative to #1210 where we used a mathematical interval notation to represent the range. That implementation brings in the complexity of parsing/regexing and reinventing what Guava has implemented. Maybe we should start with this simpler alternative and could introduce a `range` String to the `RangeSpec` later, which should be a small PR on top of the existing code to parse from String -> Range. 

This alternative also gives us type safety on the integers through Jackson. The implementation has been simplified to separate the Range as a concept from any idea of indexing the node ids in the range. It borrows the term Range from guava and the JDK Stream api, it seems like the most common term for this.

The existing PortPerBroker scheme is also refactored to be implemented using the Range aware implementation, treating it as a single-range configuration.

## Config Validations

1. Ranges must not overlap, node ids must belong to exactly one specified range
    ```yaml
            nodeIdRanges:
              - name: brokers
                range:
                  startInclusive: 0
                  endExclusive: 3
              - name: controllers
                range:
                  startInclusive: 2
                  endExclusive: 4
    ```
    causes the proxy to not start with:
    > Caused by: java.lang.IllegalArgumentException: some nodeIdRanges collided (one or more node ids are present in multiple ranges): 'brokers:[0,3)' collides with 'controllers:[2,4)'
2. Ranges must have unique names
    ```yaml
            nodeIdRanges:
              - name: brokers
                range:
                  startInclusive: 0
                  endExclusive: 3
              - name: brokers
                range:
                  startInclusive: 3
                  endExclusive: 4
    ```
    causes the proxy to not start with:
    > Caused by: java.lang.IllegalArgumentException: non-unique nodeIdRange names discovered: [brokers]
3. Ranges must not be empty
    ```yaml
            nodeIdRanges:
              - name: brokers
                range:
                  startInclusive: 0
                  endExclusive: 0
    ```
    causes the proxy to not start with:
    > Caused by: java.lang.IllegalArgumentException: invalid nodeIdRange: brokers, end of range: 0 (exclusive) is before start of range: 0 (inclusive)
### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/Range.java:
```diff
@@ -0,0 +1,72 @@
+ *
+ * A Range must be non-empty, endExclusive must be greater than startInclusive
+ */
+class Range {
```
could this be a record?

robobario replied:
yes, I started out extending `AbstractSet` and backed off tying it to the Collections api. It should become a record now.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/Range.java:
```diff
@@ -0,0 +1,72 @@
+        return IntStream.range(startInclusive, endExclusive);
+    }
+
+    public boolean isEndBeforeStartOf(@NonNull Range range) {
```
```suggestion
    @VisibleForTestingh
    public boolean isEndBeforeStartOf(@NonNull Range range) {
```

robobario commented:
Good ideas Keith,

I've updated to the:

```
nodeStartPort: 9000
nodeIdRanges:
  - name: brokers
    range:
      startInclusive: 4
      endExclusive: 6
  - name: controllers
    range:
       startInclusive: 1
       endExclusive: 3
```
syntax, I agree it allows us to have flexible configuration per rangeSpec separate from how we define the range

I've converted `Range` to a record and removed the `name` knowledge from this pure mathematical part of the code. 

I also decided that the `RangeSet` wasn't really adding enough value since it was just validating the ranges were distinct before they were converted to the offset Map. So I've made the Config class responsible for validating that the ranges are distinct.

Also I added a check that the range names are unique.

SamBarker reviewed (approved):
LGTM 

a few small things but the change is about right to me.

robobario commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.java:
```diff
@@ -0,0 +1,257 @@
+        this.nodeIdToPortOffset = config.nodeIdToPortOffset;
+
+        int nodeEndPortExclusive = nodeStartPort + config.nodeIdToPortOffset.size();
+        var exclusivePorts = IntStream.range(nodeStartPort, nodeEndPortExclusive).boxed().collect(Collectors.toCollection(HashSet::new));
```
It's a bit confusing yeah, we are working with ports here. There are three things, nodeIds, offsets and ports. This is converting offsets to ports. Maybe we don't need this and it should be a `nodeIdToPort` map and when calculation the exclusivePorts we can take `nodeIdToPort.values() + [bootstrapAddress.port()]`

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.java:
```diff
@@ -0,0 +1,257 @@
+     * @param startInclusive the (inclusive) initial value
+     * @param endExclusive the exclusive upper bound
+     */
+    public record IntRangeSpec(@JsonInclude(JsonInclude.Include.ALWAYS) @JsonProperty(required = true) int startInclusive,
```
Why the need for the separate Spec object?  Why can't we just use Range?

robobario replied:
One reason is we can delay the validation so that if an invalid range is specified the error can be associated with a range name and present a more useful message identifying the problem range.

k-wall commented on kroxylicious-runtime/src/main/resources/META-INF/services/io.kroxylicious.proxy.clusternetworkaddressconfigprovider.ClusterNetworkAddressConfigProviderContributor:
```diff
@@ -1,2 +1,3 @@
 io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider.PortPerBrokerContributor
```
Should we deprecate port per broker?

k-wall replied:
I think I'd be happy to let the new feature bake for a while in the community and defer a decision on deprecation.  
 I see this work as an improvement, but not really a desired end state.

SamBarker commented on kroxylicious-integration-test-support/src/main/java/io/kroxylicious/test/tester/KroxyliciousConfigUtils.java:
```diff
@@ -11,6 +11,7 @@
 import io.kroxylicious.proxy.config.ConfigurationBuilder;
 import io.kroxylicious.proxy.config.VirtualClusterBuilder;
 import io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider.PortPerBrokerClusterNetworkAddressConfigProvider;
+import io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider.RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.RangeAwarePortPerNodeClusterNetworkAddressConfigProviderConfig;
```
I still can't come up with a better name...

k-wall reviewed (approved):
(non-blocking) it'd be good if you can squash a couple of the low hanging sonar issues.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/Range.java:
```diff
@@ -0,0 +1,71 @@
+     * @throws IllegalArgumentException if end is before start
+     */
+    public Range {
+        if (endExclusive <= startInclusive) {
+            throw new IllegalArgumentException(
+                    "end of range: " + endExclusive + " (exclusive) is before start of range: " + startInclusive + " (inclusive)");
+        }
```
+0.5 
this Range class is modelling the mathematical concept, so I wouldn't overload it with domain logic without renaming it.  if I were going to enforce the 65535 validation (i'm not sure I'd bother), I would do it do it in the caller (RangeAwarePortPerNodeClusterNetworkAddress) rather than here.

k-wall replied:
I see you did just that already.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.java:
```diff
@@ -0,0 +1,269 @@
+                }
+            }
+            if (!collisions.isEmpty()) {
+                throw new IllegalArgumentException("some nodeIdRanges collided (one or more node ids are present in the following ranges): "
```
```suggestion
                throw new IllegalArgumentException("some nodeIdRanges collided (one or more node ids are duplicated in the following ranges): "
```

k-wall commented on kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProviderTest.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Test
+    void rangesMustBeNonEmpty() {
+        assertThatThrownBy(() -> getConfig(List.of()))
```
addresses sonar low hanging fruit (i hope)
```suggestion
        var empty = List.of();
        assertThatThrownBy(() -> getConfig(empty))
```

robobario replied:
Thanks, thought I got them all, but no!
